Stepper motor code rework

alexproskurin opened the pull request.

noaht11 committed:
333 Hz signal on pin 13 (PB27)
Switched to pin 6 for motor pulse output (PC24)

Commit:
Switched from macros to variables - achieved tighter timing on the pwm signal

Commit:
Working acceleration code

Commit:
Have functions running a single motor until a simulated number of encoder counts

Commit:
working trapezoidal profile

Commit:
2 axis setup and running

Commit:
Implemented direction to move relative function
Direction logic is working

Commit:
Added typedef to structs

Commit:
Added an option to specify max velocity for trap move - needed for homing

Commit:
Added homing routine - needs to be tested

Commit:
Created modules and moves them to lib/

Commit:
Moved interrupt setup functions
Fixed encoder ISr mistake
Fixed typo
Fixed encoder ISR to subtract when moving in negative direction

noaht11 commented on firmware/src/main.cpp:
```diff
@@ -1,16 +1,32 @@
-    digitalWrite(LED_BUILTIN, HIGH);
-    delay(1000);
-    digitalWrite(LED_BUILTIN, LOW);
-}
```
Leave this in for now I think, it's a convenient way to know the code is running

noaht11 commented on firmware/lib/Axis/src/Axis.h:
```diff
@@ -0,0 +1,63 @@
+#define VEL_MIN_Y               100
+#define VEL_MAX_Y               1000
+#define STEP_PIN_Y              22 // need to know the register for fast operation pin 5 - PB26 
+#define DIR_PIN_Y               23 // need to know the register for fast operation pin 6 - PA14
```
Comments don't match the pin numbers

noaht11 commented on firmware/lib/Axis/src/Axis.h:
```diff
@@ -0,0 +1,63 @@
+void TC3_Handler(void);
+
+// isr for operating x axis motor acceleration 
+void TC4_Handler(void);
```
Don't need to declare these in the .h file if no-one else is going to use them. I think all 4 methods above are only used internally in Axis.cpp

noaht11 commented on firmware/lib/Axis/src/Axis.cpp:
```diff
@@ -0,0 +1,264 @@
+void isr_limit_switch_y()
+{   
+    static unsigned long last_interrupt_time = 0;
+    unsigned long interrupt_time = millis();
```
Use `uint64_t`

noaht11 commented on firmware/lib/Axis/src/Axis.cpp:
```diff
@@ -0,0 +1,264 @@
+void isr_limit_switch_x()
+{   
+    static unsigned long last_interrupt_time = 0;
+    unsigned long interrupt_time = millis();
```
Use ~`uint64_t`~ `uint32_t`.
`millis()` actually returns a `uint32_t`

noaht11 commented:
Can you delete all the print statements, gonna clog up the serial port

noaht11 commented on firmware/lib/Axis/src/Axis.h:
```diff
@@ -0,0 +1,63 @@
+#define ENCODER_PIN_A_X         7
+#define ENCODER_PIN_B_X         8
+#define LIMIT_SW_HOME_PIN_X     9
+#define LIMIT_SW_FAR_PIN_X      10
```
It seems like most if not all of these defines are only used in the Axis.cpp file. If that's the case they don't need to be defined in the header and can be moved to the cpp file.
Same for the Y defines.
If you think we will want to use them somewhere else in the future you can leave them here, but at least for things like pin numbers I don't imagine they will get used anywhere else.

noaht11 commented on firmware/lib/Axis/src/Axis.h:
```diff
@@ -0,0 +1,63 @@
+#define AXIS_H
+
+#include "Debug.h"
+#include "Kinematics.h"
```
Only include in the header file, what's needed by the header file. If it's needed by the cpp file you can include it directly in the cpp file. So I think both of these includes can be moved to Axis.cpp

noaht11 commented on firmware/lib/Axis/src/Axis.cpp:
```diff
@@ -0,0 +1,264 @@
+    Axis axis_x_temp = {DIR_PIN_X, STEP_PIN_X, ACCEL_X, VEL_MAX_X, VEL_MIN_X, 
+                        {0, 0, 0}, 0, encoder_x, ls_home, ls_far_from_home,
+                        TC1, 0, 1, TC3_IRQn, TC4_IRQn};
+    return axis_x_temp;
```
This method is basically identical to the y setup one. Can we do something like declare a:
```c
typedef struct {
    uint32_t encoderPinA, encoderPinB;
    uint32_t limitSwHomePin;
    // etc.
} AxisConfig;
```
Then have one method that you pass a `AxisConfig *` to

alexproskurin replied:
Cool, will look into making it more concise

noaht11 commented on firmware/lib/Axis/src/Axis.h:
```diff
@@ -0,0 +1,63 @@
+
+void setup_ls_interrupts();
+
+#endif
```
Minor thing but standard practice is to put a comment after an `#endif` for what it's ending (especially when it's really far away from the `#if`):
```c
#endif // AXIS_H
```

noaht11 commented on firmware/lib/Kinematics/src/Kinematics.h:
```diff
@@ -0,0 +1,45 @@
+    IRQn_Type isr_accel;
+    int dir;
+    bool homing;
+} Axis;
```
Any reason this is defined in `Kinematics.h` rather than `Axis.h`?

noaht11 commented on firmware/lib/Kinematics/src/Kinematics.h:
```diff
@@ -0,0 +1,45 @@
+typedef struct Axis
+{
+    int dir_pin;
+    int step_pin;
```
Any reason these are `int`s rather than `uint32_t`s. `pinMode` and `digitalWrite` both take `uint32_t` as an argument

noaht11 commented on firmware/lib/Kinematics/src/Kinematics.h:
```diff
@@ -0,0 +1,45 @@
+
+enum status{pressed, depressed};
+enum direction{positive, negative};
+enum segment{accelerate, hold, decelerate};
```
Should typedef all these enums:
```c
typedef enum { PRESSED, DEPRESSED } Status;
// etc.
```
Also think it would be worthwhile using UPPER_CASE_UNDERSCORES style so they don't look like local variables.
And use PascalCase for the name of the enum.

After typedefing them, you can use the name of the enum everywhere instead of int, so it's clear what you're actually expecting.

noaht11 commented on firmware/lib/Kinematics/src/Kinematics.h:
```diff
@@ -0,0 +1,45 @@
+    uint32_t vel_max;
+    uint32_t vel;
+    int32_t vel_profile_cur_trap[3]; // defined in counts accelerate, hold, decelerate
+    int tragectory_segment;
```
After typedefing the enums, replace `int` with `Segment`. Same for the other ones.

alexproskurin replied:
so, this isnt actually an enum, since i didnt think i could ++ enums i just kept it as int, but i compare it to enum for readability

noaht11 commented on firmware/lib/Axis/src/Axis.cpp:
```diff
@@ -0,0 +1,264 @@
+void isr_encoder_y()
+{
+    static unsigned long last_interrupt_time = 0;
+    unsigned long interrupt_time = millis();
```
Use ~`uint64_t`~ `uint32_t`.
`millis()` actually returns a `uint32_t`

noaht11 commented on firmware/lib/Axis/src/Axis.cpp:
```diff
@@ -0,0 +1,264 @@
+#include "Axis.h"
+
+// -------------------------------- X AXIS SETUP -------------------------------- //
+Axis setup_axis_x()
```
Given that all the pinMode configurations happen here, why not attach the interrupts as well?

alexproskurin replied:
Yeah, i kinda move things into modules super quickly and need to give more thought to the location of different things

noaht11 commented on firmware/lib/Axis/src/Axis.cpp:
```diff
@@ -0,0 +1,264 @@
+{   
+    // print("Desired count: ", axis_x.encoder.desired);
+    // print("Current count: ", axis_x.encoder.current);
+    TC_GetStatus(TC1, 0);
```
Can you comment what TC_GetStatus does. I'm assuming it clears/acks the interrupt?

noaht11 commented on firmware/lib/Axis/src/Axis.cpp:
```diff
@@ -0,0 +1,264 @@
+            return;
+        }
+        else {
+            axis_x.tragectory_segment++;
```
Are you incrementing an enum...I would classify that as relatively dangerous.
Especially since you didn't define numerical values for the enum constants.
I would say it's better to just hardcode what the next value is

alexproskurin replied:
Not incrementing enum here, that's why segment is defined as int

Commit:
Changed axis config to be a function
Moved stuff from kinematics into axis
Moved attach interrupt to axis setup
Removed unimplemented setup code
Fixed wrong channel assignment
Working stoppping code
Addresssed large data structure passing

noaht11 commented on firmware/lib/Axis/src/Axis.cpp:
```diff
@@ -0,0 +1,317 @@
+    // TC1 channel 0, the IRQ for that channel and the desired frequency
+    pinMode(axis_config->encoder_pin_a, INPUT);
+    pinMode(axis_config->encoder_pin_b, INPUT);
+    Encoder encoder_x = {axis_config->encoder_pin_a, 10, 0};
```
This no longer specific to x right? so just `encoder`?

noaht11 commented on firmware/lib/Axis/src/Axis.cpp:
```diff
@@ -0,0 +1,317 @@
+AxisConfig axis_x_config = {DIR_PIN_X, STEP_PIN_X, ENCODER_PIN_A_X, ENCODER_PIN_B_X, LIMIT_SW_HOME_PIN_X, LIMIT_SW_FAR_PIN_X,
+                            ACCEL_X, VEL_MAX_X, VEL_MIN_X, TC1, 0, 1, TC3_IRQn, TC4_IRQn, isr_encoder_x, isr_limit_switch_x};
+
+Axis setup_axis(AxisConfig *axis_config)
```
I guess I missed this on the first round, but generally speaking you don't want to return large structs from functions because they're being passed around by value. The standard practice would be to do something like this:
```c
void setup_axis(AxisConfig *axis_config, Axis *axis)
{
    axis->blah = blah;
    axis->blah2 = blah2;
    ...
}

AxisConfig axis_x_config = ...;
Axis axis_x;

void some_init_method()
{
    setup_axis(&axis_x_config, &axis_x);
}
```

noaht11 commented on firmware/lib/Axis/src/Axis.cpp:
```diff
@@ -0,0 +1,317 @@
+    return axis_x_temp;
+}
+
+Axis axis_x = setup_axis(&axis_x_config);
```
Don't love the idea of calling functions just in the global scope outside of any other function.
Especially ones that have side effects like setting pinModes and attaching interrupts.

Can we move these to some init function that can be called from main or just call them from main?

noaht11 commented on firmware/lib/Axis/src/Axis.cpp:
```diff
@@ -0,0 +1,317 @@
+                            ACCEL_Y, VEL_MAX_Y, VEL_MIN_Y, TC0, 0, 1, TC0_IRQn, TC1_IRQn, isr_encoder_y, isr_limit_switch_y};
+
+
+Axis axis_y = setup_axis(&axis_y_config);
```
See comments about axis_x (applies here as well)

Commit:
Moved .cpp and .h files into Gantry folder

Commit:
Added stop axis and waiting for count_accel function
Added reset function and made some of the functions static that shouldn't be called outside of the file

noaht11 committed:
Implemented Kinematics math

noaht11 committed:
Replaced debug prints with a macro

noaht11 committed:
Disabled initial motor run

noaht11 committed:
Extracted movement code into separate files from Axis

Commit:
Working homing routine
